test/e2e/autoscale_test.go

```diff
@@ -245,23 +189,147 @@ func TestAutoscaleUpDownUp(t *testing.T) {
+			// ctx.logger.Infof( /* DO NOT SUBMIT */ "sending request to %v", fmt.Sprintf("http://%s", ctx.domain))
+			for {
+				select {
+				case <-done:
+					return nil
+				default:
+					mux.Lock()
+					totalRequests++
+					mux.Unlock()
+					res, err := client.Do(req)
+					if err != nil {
+						ctx.logger.Debugf("error making request %v", err)
+						continue
+					}
+
+					if res.StatusCode != http.StatusOK {
+						ctx.logger.Debugf("non 200 response %v", res.StatusCode)
+						ctx.logger.Debugf("response headers: %v", res.Header)
+						ctx.logger.Debugf("response body: %v", res.Body)
+						continue
+					}
+					mux.Lock()
+					successfulRequests++
+					mux.Unlock()
+				}
+			}
+		})
+	}
+
+	ctx.logger.Infof("Waiting for all requests to complete.")
+	if err := group.Wait(); err != nil {
+		return fmt.Errorf("error making requests for scale up: %v", err)
+	}
+
+	if successfulRequests != totalRequests {
+		return fmt.Errorf("error making requests for scale up. Got %d successful requests. Wanted %d",
+			successfulRequests, totalRequests)
+	}
+
+	ctx.logger.Infof("All %d requests succeeded.", totalRequests)
+	return nil
+}
+
+func bounceDeployment(ctx *testContext, name string) {
+	ctx.logger.Infof("Bouncing deployment %q", name)
+	dep, err := ctx.clients.KubeClient.Kube.AppsV1().Deployments("knative-serving").Get(name, metav1.GetOptions{})
 	if err != nil {
-		t.Fatalf("Error during final scale up: %v", err)
+		ctx.t.Fatalf("Unable to get %v deployment: %v", name, err)
 	}
-	err = test.WaitForDeploymentState(
-		clients.KubeClient,
-		deploymentName,
-		isDeploymentScaledUp(),
-		"DeploymentScaledUp",
-		2*time.Minute)
+	selector := labels.FormatLabels(dep.Spec.Selector.MatchLabels)
+	originalPods, err := ctx.clients.KubeClient.Kube.CoreV1().Pods("knative-serving").List(metav1.ListOptions{
+		LabelSelector: selector,
+	})
+	if err != nil {
+		ctx.t.Fatalf("Unable to list %v pods: %v", name, err)
+	}
+	if len(originalPods.Items) != 1 {
+		ctx.t.Fatalf("Expected 1 pod. Found %v.", len(originalPods.Items))
+	}
+	originalController := originalPods.Items[0].Name
+	err = ctx.clients.KubeClient.Kube.CoreV1().Pods("knative-serving").Delete(originalController, nil)
+	if err != nil {
+		ctx.t.Fatalf("Unable to delete controller: %v", err)
+	}
+	err = wait.PollImmediate(time.Second, 2*time.Minute, func() (bool, error) {
+		newPods, err := ctx.clients.KubeClient.Kube.CoreV1().Pods("knative-serving").List(metav1.ListOptions{
+			LabelSelector: selector,
+		})
+		if err != nil {
+			return false, err
+		}
+		if len(newPods.Items) != 1 {
+			return false, nil
+		}
+		if newPods.Items[0].Name == originalController {
+			return false, nil
+		}
+		return true, nil
+	})
 	if err != nil {
-		logger.Fatalf(`Unable to observe the Deployment named %s scaling
-			   up. %s`, deploymentName, err)
+		ctx.t.Fatalf("Error verifying deployment is bounced: %v", err)
 	}
+	ctx.logger.Infof("Deployment %q has been bounced.", name)
+	time.Sleep(10 * time.Second)
```

Comment:
Should we wait for the pod to become ready? What else is this sleep waiting for?

Reply:
In the case of the autoscaler, allowing for the pods to reconnect.  That's not represented in pod readiness, so I just sleep.

Reply:
can you add a comment here on why there is a sleep? Also, if there is better solution or suggestion, we should create an issue and link it here to remove the sleep.

Reply:
I've tested this again without the sleep, and it passes.  So I've just removed this.